agent/azure-monitor-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/utils/ResourceParser.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * ApplicationInsights-Java
+ * Copyright (c) Microsoft Corporation
+ * All rights reserved.
+ *
+ * MIT License
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of this
+ * software and associated documentation files (the ""Software""), to deal in the Software
+ * without restriction, including without limitation the rights to use, copy, modify, merge,
+ * publish, distribute, sublicense, and/or sell copies of the Software, and to permit
+ * persons to whom the Software is furnished to do so, subject to the following conditions:
+ * The above copyright notice and this permission notice shall be included in all copies or
+ * substantial portions of the Software.
+ * THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+ * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
+ * PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE
+ * FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+ * OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ */
+
+package com.azure.monitor.opentelemetry.exporter.implementation.utils;
+
+import com.azure.monitor.opentelemetry.exporter.implementation.builders.AbstractTelemetryBuilder;
+import com.azure.monitor.opentelemetry.exporter.implementation.models.ContextTagKeys;
+import io.opentelemetry.sdk.resources.Resource;
+import io.opentelemetry.semconv.resource.attributes.ResourceAttributes;
+import java.util.HashMap;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+public final class ResourceParser {
+
+  public static void updateRoleNameAndInstance(
+      AbstractTelemetryBuilder builder, Resource resource) {
+    Map<ContextTagKeys, String> resourceMap = ResourceParser.parseRoleNameAndInstance(resource);
+    if (resourceMap != null && !resourceMap.isEmpty()) {
+      resourceMap.forEach(
+          (key, value) -> {
+            String stringKey = key.toString();
+            if (ContextTagKeys.AI_CLOUD_ROLE.toString().equals(stringKey)
+                || ContextTagKeys.AI_CLOUD_ROLE_INSTANCE.toString().equals(stringKey)) {
+              builder.addTag(stringKey, value);
+            }
+          });
+    }
+  }
+
+  // visible for test
+  @Nullable
+  static Map<ContextTagKeys, String> parseRoleNameAndInstance(@Nullable Resource resource) {
+    if (resource == null) {
+      return null;
+    }
+    String serviceName = resource.getAttribute(ResourceAttributes.SERVICE_NAME);
+    if (serviceName == null || serviceName.isEmpty()) {
+      serviceName = Resource.getDefault().getAttribute(ResourceAttributes.SERVICE_NAME);
```

Comment:
what is the default service_name? I'm guessing (but not sure) that we don't want to send any rolename if none is configured (the U/X defaults the app map to the name of the AI Portal Resource in this case)

Reply:
"unknown_service:java". upstream has a bug there. it should not have ":java". dotnet uses "unknown_service". service_name is mandatory and will never be null nor empty.

Reply:
should I submit a PR to fix [it](https://github.com/open-telemetry/opentelemetry-java/blob/d868eff3e1d63a7d46f5866d676570cc2f2be84a/sdk/common/src/main/java/io/opentelemetry/sdk/resources/Resource.java#:~:text=private%20static%20final%20Resource%20MANDATORY,SERVICE_NAME%2C%20%22unknown_service%3Ajava%22)?

Reply:
I'm not sure it's a bug, check out the spec:

https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/semantic_conventions/README.md#service